### DISCUSSION TITLE
Parsing XML with Mojo::DOM: how to get nested nodes?

### DISCUSSION THREAD

**Discussion by fluca1978** [mojolicious/mojo#2166] [Q&A] [answered]
I'm not understanding well how to parse an XML file with ~Mojo::DOM`. I use a path to get to the nodes that are nested, so for example with `XML::LibXML` I would do:
```
my @nodes = $xml->findnodes( q{xocs:doc/xocs:item/item/bibrecord/head/author-group} );
```

that I tried to convert into a selector chain in `Mojo::DOM` but the following is not working:

```
my @nodes = $dom->find( q{xocs:doc > xocs:item > item > bibrecord > head > author-group} );
```

Surely I'm not understanding how the DOM is working, but is the above possible?

**Comment by kraih** [comment]
`Mojo::DOM` uses CSS selectors instead of XPath. There's probably CPAN modules that do the conversion for you.

**Comment by fluca1978** [comment]
> `Mojo::DOM` uses CSS selectors instead of XPath. There's probably CPAN modules that do the conversion for you.

Yes, and the CSS selector I placed is correct.
After a few tests, it seems that `Mojo::DOM` is not able to handle tags with a namespace, so `xocs:doc` is not handled, while if I change in my XML document the tag to `xocs-doc` it is parsed correctly.
Is there a trick I'm not aware of to let `Mojo::DOM` handle tag names with namespaces?

**Reply by kraih** [reply]
No, your selector is very much incorrect. What you've written as `xocs:item` is a selector for an `<xocs>` tag with an undefined  `item` pseudo-class.

**Comment by Akron** [comment, marked as answer]
There are two ways to deal with namespaces in Mojo::DOM::CSS. Either you define the namespace and use it with `|` [as described here](https://metacpan.org/pod/Mojo::DOM::CSS#A%7CE) - or you ignore the namespace and escape the colon (that way the namespace prefix is treated as being part of the tag).

```perl
use Mojo::Base -strict;
use Mojo::DOM;

my $file = join("\n",<DATA>);

my $dom = Mojo::DOM->new->xml(1)->parse($file);

$dom->find( q{yadda|doc > item}, yadda => 'xocs-ns' )->each(sub { say $_->text });

$dom->find( q{xocs\:doc > item} )->each(sub { say $_->text });

__DATA__

<xocs:doc xmlns:xocs="xocs-ns">
  <item>hello</item>
  <item>world</item>
</xocs:doc>

**Comment by fluca1978** [comment]
@Akron thanks, I was figuring out the error in my namespace selector, so I have to rewrite it as:

```
$dom->find( q{xocs|doc > xocs|item > item > bibrecord > head > author-group > author},
			 xocs => q{http://www.foo.com/xml/xocs/dtd} )
```

that's probably the best solution.

**Comment by fluca1978** [comment]
> No, your selector is very much incorrect. What you've written as `xocs:item` is a selector for an `<xocs>` tag with an undefined `item` pseudo-class.

Thanks, I figured out I need to use the `|`.
